### PR TITLE
Room state

### PR DESCRIPTION
**Pull request by fancycade**
We should merge #13 before this.

This PR addresses #7 and #11.

### CODE REVIEW

**Comment by niklaslong** [comment]
Looks like #13 introduced some conflicts, @fancycade could you rebase?

**Comment by fancycade** [comment]
Oops didnt realize the log would be that screwed up. Good thing this is an experiment :rofl:

File: lib/matrix_client.ex [hunk 1]
```diff
@@ -91,10 +91,22 @@ defmodule MatrixClient do
 
     handle_result(
       Client.joined_rooms(url, token),
-      fn body -> {:ok, Map.get(body, "joined_rooms")} end
+      fn body -> {:ok, joined_rooms_helper(session, body["joined_rooms"])} end
     )
   end
 
+  defp joined_rooms_helper(session, room_ids) do
```
Comment (niklaslong) [inline lib/matrix_client.ex]: How about something more descriptive, perhaps `map_joined_rooms` or `list_joined_rooms`?

File: lib/matrix_client.ex [hunk 2]
```diff
@@ -146,11 +166,51 @@ defmodule MatrixClient do
...
+  def prev_room_messages(pid, room_id) do
+    get_room_messages(pid, room_id, "b")
+  end
+
+  defp get_room_messages(pid, room_id, direction) do
+    {:ok, url} = Session.get(pid, :url)
+    {:ok, token} = Session.get(pid, :token)
+    aliases = Session.get_aliases(pid)
+
+    new_id =
```
Comment (niklaslong) [inline lib/matrix_client.ex:190]: As an aside, I'm seeing the following block a lot—refactor opportunity in future.

File: lib/matrix_client/session.ex [hunk 3]
```diff
@@ -5,7 +5,9 @@ defmodule MatrixClient.Session do
   Starts a new session with a base_url.
   """
   def start_link(url) do
-    Agent.start_link(fn -> %{url: url, rooms: %{}, invites: %{}} end)
+    Agent.start_link(fn ->
+      %{url: url, rooms: %{}, invites: %{}, aliases: %{ids: %{}, aliases: %{}}}
```
Comment (niklaslong) [inline lib/matrix_client/session.ex]: What are the `ids` in the `aliases` map?

File: lib/matrix_client/session.ex [hunk 4]
```diff
@@ -120,8 +154,40 @@ defmodule MatrixClient.Session do
...
+  def sync_alias({room_id, room_data}, aliases) do
+    %{"timeline" => %{"events" => events}} = room_data
+
+    case filter_alias_event(events) do
+      [ae] -> alias_helper_ids(aliases, room_id, ae["content"]["alias"])
+      _ -> aliases
+    end
+  end
+
+  defp alias_helper_ids(aliases, room_id, a) do
```
Comment (niklaslong) [inline lib/matrix_client/session.ex]: Not sure what these functions are doing just by reading the name? In general I'm not a big fan of using `helper` as a descriptor as it's so vague. Thoughts?

**Review by niklaslong — CHANGES_REQUESTED** [review]
Great work—I've left a few comments. Once this is merged I'll review the project and its status in relation to the SDK.

**Comment by fancycade** [comment]
@niklaslong Updated based on feedback.

Changed any helper names or refactored.

Moved ids into its own map. The reason for it  is that internally there are times to translate an alias to an id, and then the alias map is to translate an id to an alias (typically for external reasons like listing joined rooms). Can't decide if the naming is backwards or not :laughing:

**Review by niklaslong — APPROVED** [review]

**Pull request merged**